consensus/istanbul/backend/backend.go

```diff
@@ -516,6 +488,40 @@ func (sb *Backend) getValidators(number uint64, hash common.Hash) istanbul.Valid
 	return snap.ValSet
 }
 
+// validatorRandomnessAtBlockNumber calls into the EVM to get the randomness to use in proposer ordering at a given block.
+func (sb *Backend) validatorRandomnessAtBlockNumber(number uint64, hash common.Hash) (common.Hash, error) {
+	lastBlockInPreviousEpoch := number
+	if number > 0 {
+		lastBlockInPreviousEpoch = number - istanbul.GetNumberWithinEpoch(number, sb.config.Epoch)
+	}
+	header := sb.chain.GetHeader(hash, lastBlockInPreviousEpoch)
```

Comment:
so looks like we will not be able to verify blocks without having already fetched the last header of that epoch? I feel it would be better to have all of the info we need to verify a block be self-contained in the latest block.

Reply:
That would certainly be better, but we do need to pull those block headers to construct the snapshot, so in reality this is not a new requirement. What is new is pulling state at that point. If we were to put the randomness in the header that would work, but I am not sure whether the tradeoffs would be appropriate.